pkg/istio-agent/xds_proxy.go

```diff
@@ -367,29 +397,73 @@ func (p *XdsProxy) handleUpstreamResponse(con *ProxyConnection) {
 					TypeUrl:       v3.NameTableType,
 					ResponseNonce: resp.Nonce,
 				}
+			case v3.ExtensionConfigurationType:
+				// intercepts ECDS, replace Wasm remote load with local file path.
+				if dr.wasmExtensionConverted {
+					forwardToEnvoy(con, resp)
+				} else {
+					p.ecdsUpdateChan <- discoveryResponse{
+						resp:                   resp,
+						wasmExtensionConverted: false,
+					}
+				}
 			default:
-				// TODO: Validate the known type urls before forwarding them to Envoy.
-				if err := sendDownstreamWithTimeout(con.downstream, resp); err != nil {
-					proxyLog.Errorf("downstream send error: %v", err)
-					// we cannot return partial error and hope to restart just the downstream
-					// as we are blindly proxying req/responses. For now, the best course of action
-					// is to terminate upstream connection as well and restart afresh.
-					con.downstreamError <- err
-					return
+				forwardToEnvoy(con, resp)
+			}
+		case <-con.stopChan:
+			return
+		}
+	}
+}
+
+func (p *XdsProxy) handleUpstreamECDSResponse(con *ProxyConnection) {
+	// TODO(bianpengyuan): Add metrics about ecds handling.
+	for {
+		select {
+		case discoveryResp := <-p.ecdsUpdateChan:
+			sendNack := wasm.MaybeConvertWasmExtensionConfig(discoveryResp.resp.Resources, p.wasmCache)
+			if sendNack {
+				con.requestsChan <- &discovery.DiscoveryRequest{
+					VersionInfo:   p.ecdsLastAckVersion,
```

Comment:
why is it  p.ecdsLastAckVersion and not discoveryResp.resp.VersionInfo? Shouldn't we NACK the version that triggered the error?

Reply:
From xds spec doc, it appears to be the version number should be the previous version and error details? https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#nack

Reply:
Ah, good catch